### PR TITLE
Add 1D sinusoid analytic solution to ScalarAdvection

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

To ScalarAdvection system,
* add 1D sinusoid analytic data
* add 1D sinusoid analytic solution

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  // check if analytic solution equals to analytic data at t=0
+  ScalarAdvection::AnalyticData::Sinusoid initial_data;
+  ScalarAdvection::Solutions::Sinusoid solution;
+  u_sol = solution.u<DataVector>(x, 0.0);
+  u_test = initial_data.u<DataVector>(x);
+  CHECK_ITERABLE_APPROX(u_sol, u_test);
+
+  // compare u and du_dt values from analytic solution with python
+  // implementation
+  double t{0.3};
```
Comment (yoonso0-0) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp:37]: @nilsdeppe This value is completely arbitrary. I was not sure if the analytic solution should be tested for multiple values of random `t` ?

File: src/PointwiseFunctions/AnalyticData/ScalarAdvection/Sinusoid.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * \f{align}{
+ *   u(x, 0) = \sin \pi x
+ * \f}
+ *
+ * At future times the analytic solution can be found by simply advecting the
+ * initial data (sine wave) with the velocity of unity:
+ *
+ * \f{align}{
+ *   u(x,t)=\sin \pi(x-t)
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticData/ScalarAdvection/Sinusoid.hpp:37]: since the solution is quite easy, why not just make this an `AnalyticSolution`? I guess that's because it depends on the chosen velocity? If so, maybe state what the solution would be as a function of velocity or maybe reword "If the advection velocity is unity, then at..."?

File: tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  test_serialization(ScalarAdvection::AnalyticData::Sinusoid{});
+  const auto periodic =
+      TestHelpers::test_creation<ScalarAdvection::AnalyticData::Sinusoid>("");
+  CHECK(periodic == ScalarAdvection::AnalyticData::Sinusoid{});
+  test_move_semantics(ScalarAdvection::AnalyticData::Sinusoid{},
+                      ScalarAdvection::AnalyticData::Sinusoid{});
+
+  pypp::check_with_random_values<1>(
+      &ScalarAdvectionSinusoidProxy::variables, ScalarAdvectionSinusoidProxy{},
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp:39]: can you add a serialization test to make sure the `pup` function works. While in this case it's pretty trivial, bugs related to `pup` functions are pretty annoying because the values change seemingly randomly when running the executable

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp (hunk 1 above)
Comment (nilsdeppe) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp:37]: you can generate a random time if you prefer. We generally like that since it prevents magic values being accidentally chosen

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  Sinusoid() = default;
+  Sinusoid(const Sinusoid&) noexcept = default;
+  Sinusoid& operator=(const Sinusoid&) noexcept = default;
+  Sinusoid(Sinusoid&&) noexcept = default;
+  Sinusoid& operator=(Sinusoid&&) noexcept = default;
+  ~Sinusoid() noexcept = default;
+
+  template <typename T>
+  Scalar<T> u(const tnsr::I<T, 1>& x, double t) const noexcept;
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:45]: Burgers is a bit dated in how it's structured. We don't need or use functions like `u` and `du_dt` anymore, it all goes through the `variables` functions

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  template <typename T>
+  Scalar<T> du_dt(const tnsr::I<T, 1>& x, double t) const noexcept;
+
+  tuples::TaggedTuple<Tags::U> variables(
+      const tnsr::I<DataVector, 1>& x, double t,
+      tmpl::list<Tags::U> /*meta*/) const noexcept;
+
+  tuples::TaggedTuple<::Tags::dt<Tags::U>> variables(
+      const tnsr::I<DataVector, 1>& x, double t,
+      tmpl::list<::Tags::dt<Tags::U>> /*meta*/) const noexcept;
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:56]: we no longer need access to the time derivatives, so you can just remove this function. We needed these before we had self-starting Adams-Bashforth time stepping

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+namespace Solutions {
+/*!
+ * \brief An 1D sinusoidal wave advecting with speed 1.0.
+ *
+ * \f{align*}
+ *  u(x,t) = \sin \pi(x-t)
+ * \f}
+ *
+ */
+class Sinusoid : public MarkAsAnalyticSolution {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:32]: I'm not sure why you have both an analytic data and an analytic solution of the sinusoid. Why not just have the solution?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Systems/ScalarAdvection/Tags.hpp"
+#include "Framework/CheckWithRandomValues.hpp"
+#include "Framework/SetupLocalPythonEnvironment.hpp"
+#include "PointwiseFunctions/AnalyticData/ScalarAdvection/Sinusoid.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp"
+
+namespace {
+void test_sinusoid_solution(const gsl::not_null<std::mt19937*> generator,
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp:15]: I'd recommend looking at something like `Test_TovStar.cpp` for an idea of how we do tests nowadays. The Burgers code is very dated. The things that seem to be missing are serialization tests (and even TOV doesn't do that super well, it's better to feed the deserialized object through the full tests), copy and move semantics, equivalence operators, and creating from options. `Test_ShuOsherTube.cpp` might be the most modern example I can think of off the top of my head.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp (hunk 6 above)
Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:32]: Oh. I thought that analytic data and solution are sort of different code entities and should be written separately to be used. Surely in this case it will be possible to use analytic solution with t=0 as an analytic data..

Do you mean that when specifying EvolutionMetavars in EvolveScalarAdvectionFwd.hpp, it is sufficient to just add `Solutions::Sinusoid` rather than adding `AnalyticData::Sinusoid` as well?

Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:32]: I just looked through other `Evolve##Fwd.hpp` files and made a hypothesis..
1. If initial data + exact solution exists (e.g. IsentropicVortex) -> implement as **analytic solution** and use it as EvolutionMetavars
2. If initial data can be specifed but there's no exact solution (e.g. KH Instability) -> implement as **analytic data** and use it as EvolutionMetavars

and perhaps there will be no observable field such as `Error(Velocity_x)` in the .h5 volume file for the case 2? Was my guess correct?

File: src/PointwiseFunctions/AnalyticData/ScalarAdvection/Sinusoid.hpp (hunk 2 above)
Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticData/ScalarAdvection/Sinusoid.hpp:37]: Please see my comment below!

File: tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp (hunk 3 above)
Comment (yoonso0-0) [reply tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp:39]: Can you explain about that test and the `pup` function? To be honest `pup.h`, `class er`, `void pup(PUP::er& p) override` are something I've been merely copy&pasting without further thoughts 😓

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp (hunk 7 above)
Comment (yoonso0-0) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Test_Sinusoid.cpp:15]: Okay! I'll take a look at those.

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp (hunk 6 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:32]: Yep, that's right. If you're setting the variables to a time-dependent function that satisfies the PDEs (an analytic solution), you don't need an analytic data of the same thing since you can just evaluate the solution at `t=0`

File: tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp (hunk 3 above)
Comment (nilsdeppe) [reply tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp:39]: `pup` stands for `pack and unpack`. This takes a (possibly complicated) data structure (class) and turns it into a string of bytes (eg std::vector<char>) so that the class can be sent from one processor to another over a network. The packing part is taking the data structure and turning it into a string of bytes, while the unpacking part takes the string of bytes and restores the data structure from it

Comment (yoonso0-0) [reply tests/Unit/PointwiseFunctions/AnalyticData/ScalarAdvection/Test_Sinusoid.cpp:39]: Thank you!

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp (hunk 6 above)
Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:32]: Okay, thanks!
